### PR TITLE
Replicate sectors functionality

### PR DESCRIPTION
**Pull request by camilleldn**
This branch was manually pushed to this Heroku app for sharing with the team:
https://work-you-could-do.herokuapp.com

Left to do to complete the transfer:
- [ ] tests and CI including stylecheck
- [x] check accessibility
- [ ] actually send email on sending email
- [ ] dockerisation
- [ ] push to ECS

This pull request is here for early feedback! All feedback welcome :)

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -27,16 +27,20 @@ gem 'sdoc', '~> 0.4.0', group: :doc
 gem 'govuk_template'
 gem 'govuk_frontend_toolkit'
 
+# Call external API
+gem 'httparty'
+
 group :development, :test do
   # Call 'byebug' anywhere in the code to stop execution and get a debugger console
   gem 'byebug'
+  gem 'pry'
```
Comment (gjvis) [inline Gemfile]: Should we just use `https://github.com/rweng/pry-rails`?

File: app/assets/javascripts/report.coffee [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (gjvis) [inline app/assets/javascripts/report.coffee]: delete me :)

File: app/assets/stylesheets/report.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Report controller here.
```
Comment (gjvis) [inline app/assets/stylesheets/report.scss]: If its ok can we not have random empty scss files checked in? I know scaffolding generates them, but they're just IDE noise

File: app/controllers/report_controller.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
+class ReportController < ApplicationController
+  def select_soc_codes
+    report = Report.where(guid: params[:guid]).first
+    @keywords = report.keywords
+  end
+
+  def save_soc_codes
```
Comment (gjvis) [inline app/controllers/report_controller.rb]: can we put non-action methods at the end and make them private?

File: config/routes.rb [hunk 5]
```diff
@@ -6,8 +6,11 @@
   root 'start#index'
 
   get 'cookies' => 'start#cookies'
-  get 'select_soc_codes' => 'start#select_soc_codes'
-  get 'report' => 'start#report'
+  post 'keywords' => 'start#keywords'
+  get ':guid/select_soc_codes' => 'report#select_soc_codes'
+  post ':guid/save_soc_codes' => 'report#save_soc_codes'
```
Comment (gjvis) [inline config/routes.rb]: oh, they're action method. isn't life going to be easier if we model this with resources?

File: lib/lmi_for_all.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
+class LmiForAll
+  def initialize(soc_code)
+    @soc_code = soc_code
```
Comment (gjvis) [inline lib/lmi_for_all.rb]: Why not just pass in the `soc_code` on the lookup method? (Then we can avoid giving this class any state)

File: app/controllers/report_controller.rb (hunk 4 above)
Comment (gjvis) [inline app/controllers/report_controller.rb]: I see that this is an action method, which feels a bit weird. Can we model this with resources? I haven't really given it tell me if the answer is "lol no"

**Comment by gjvis** [comment]
Does this need a database at all? I think it there might be fewer moving parts if we did something like:

- A "report" is just the parameters entered
- All state is on the querystring
- We query the lmi4all api directly every time (potentially through a caching proxy)

File: Gemfile (hunk 1 above)
Comment (camilleldn) [inline Gemfile]: yes! will fix :)

File: app/assets/stylesheets/report.scss (hunk 3 above)
Comment (camilleldn) [inline app/assets/stylesheets/report.scss]: cool, will go through, I think there's a couple more empty files from generators still kicking around

File: app/controllers/report_controller.rb (hunk 4 above)
Comment (camilleldn) [inline app/controllers/report_controller.rb]: "lol yes" actually :) writing some tests at the moment which has driven doing exactly that :)

File: config/routes.rb (hunk 5 above)
Comment (camilleldn) [inline config/routes.rb]: yes yes :)

File: lib/lmi_for_all.rb (hunk 6 above)
Comment (camilleldn) [inline lib/lmi_for_all.rb]: :+1: currently refactoring this to the following structure:

`LmiForAll.new(lmi_client).lookup(soc_code)`

**Comment by camilleldn** [comment]
@gjvis this needs a database in the context of being able to refer back to it in the future with unique reference (the prototype for user research includes registering actions that you could refer to in the future)
If that idea goes nowhere and doesn't test well, you could be well right that database storage is unnecessary here :(

**Comment by gjvis** [comment]
If all state is on the querystring then that's covered, in fact its probably better covered as no-one else could accidentally change your report

**Comment by camilleldn** [comment]
@gjvis then share the entire URL with the work coach instead of just a unique reference?

**Comment by gjvis** [comment]
I think so – and it might simplify our lives in the immediate future?

**Comment by gjvis** [comment]
(there's no user-entered free text right?)

**Comment by camilleldn** [comment]
@gjvis beyond "it's actually more work to store in the DB" do you have other reservations about storing these?
(I'm also approaching this from the point of view of "we could run queries to see average usage, see which SOC codes don't get selected, average number of SOCs selected, which SOCs get more actions etc..." )
There's also the issue that in the "setting yourself actions" user research prototype there are user entered fields

**Comment by gjvis** [comment]
Adding state adds edge cases. This is going to need to run in more than just a lab, so we need to be treating it sort of like a "beta" (in GDS language) in terms of code quality.

If we're letting users add free text (forgot about that) then storing all state on the querystring starts to get a problematic :(

We also need to discuss with security whether we can have something publicly accessible on AWS with user entered info

**Comment by gjvis** [comment]
Aaaaand I've just realised that the search terms themselves are user entered info in anyh case. YAY :cry: 

For now, we should just do what's simplest in terms of report model vs querystring to get us into the lab.

(I do still think that saving the API results is a bit unnecessary though)

File: lib/lmi_for_all.rb (hunk 6 above)
Comment (gjvis) [inline lib/lmi_for_all.rb]: :+1: :+1:

File: config/routes.rb (hunk 5 above)
Comment (gjvis) [inline config/routes.rb]: Cool, makes total sense

**Comment by camilleldn** [comment]
@gjvis I agree that I went overboard on the API results caching :sweat_smile:

**Pull request merged**